keps/sig-scalability/0001-enhance-network-performance-benchmark-testcases.md

```diff
@@ -0,0 +1,152 @@
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+<!-- /toc -->
+
+## Summary
+
+We are proposing to add more testcases for network performance benchmarking.
+
+## Motivation
+
+As Kubernetes is increasingly being adopted by various telco cloud providers for deployments, 
+it brings additional networking requirements/scenarios to be supported.  
+With wide array of network functions being containerized, the need to analyse and 
+monitor the networking performance metrics such as latency, throughput etc has arisen.
+
+We have come across 2 test frameworks which have limited testcases for network performance benchmarking.
+
+a) E2E (test\e2e\network\networking_perf.go): This framework has only 2 testcases for measuring throughput.
+
+b) netperf (https://github.com/kubernetes/perf-tests/tree/master/network/benchmarks/netperf) : This framework 
+   measures only the throughput between client-server pods for various MSS(Maximum Segment size). 
+   Other performance metrics such as latency , loss-rate , response time are not measured.
+
+### Goals
+
+The primary goal of this proposal is to enhance the testcases to cover various 
+networking scenarios :
+
+a)  IPv4 , IPv6 , Dual-Stack.
+b)  Protocols: TCP, UDP, HTTP/HTTPS.
+c)  With services such as ClusterIP , NodePort etc.
+
+We intend to submit tests to measure the following network performance metrics for various 
+scenarios such as different services (ClusterIP/NodePort), with IPv4/IPv6 configured etc.
+
+a)  Throughput (For TCP traffic)
+b)  Loss rate (For UDP traffic)
+c)  Latency/Jitter (For UDP/HTTP/HTTPS)
+d)  Response time (For HTTP/HTTPs)
+
+Our important goal is to make sure these benchmarking testcases are run periodically,
+so that any degradation in performance is detected in early phase of feature delivery.
+
+### Non-goals
+
+Benchmarking and measurements with intermediate services like ingress controllers, load-balancers 
+are out of scope of this proposal.
+
+### Challenges and Open Questions [optional]
+
+1) Which is the suitable hosting framework for proposed tests ? 
+2) Periodic execution of these proposed tests ?
+
+
+## Proposal
+
+### Test Scenarios
+
+As captured in the link https://docs.google.com/document/d/1_760ptAHBY_c56slWhjl6PJQOrf__vQFJXkL58ewg1M/edit
+we propose to have 2 broad categories based on how client PODs can communicate with server PODs.
+
+The testcase will cover different protocols such as TCP, UDP, HTTP/HTTPS with IPv4/IPv6 addresses configured.
+
+The testcases will cover different client-to-server pod number ratio (i.e 1:N , 1:1 , N:1)
+
+### Tools
+
+We intend to use the following tools for testing in addition to iPerf/netperf already being used
+
+a)  T-Rex: T-Rex is opensource traffic generator tool made available by CISCO.
+    It will be used for measuring the UDP maximum throughput (Kpps), RTT latency (min,max,avg).
```

Comment:
you should expand on this, is Trex going to be deployed as a pod? Stateful or stateless mode?
If you only use one Trex pod, how do you "replay back" the generated traffic?

Reply:
In our simulated setup, we have tried with Trex client pod (in stateless mode) and server pod with OVS installed.
But to replay the received packets at server pod , we had to use 2 vNICs. Hence both client and server pods had to use 2 vNICs. This forced us to use multus CNI. This could become a dis-advantage if i can't use secondary network CNI. 

iperf2 also gives latency (end-to-end) , but with Trex we could measure NDR/PDR (https://trex-tgn.cisco.com/trex/doc/trex_ndr_bench_doc.html).

So we need to decide here.

Reply:
yeah, I have used Trex quite a bit in the past, it's really a great and powerful tools and I love to be used here.

For my Kubernetes testing with Trex I had to use it in Advanced Stateless mode  (something similar to this setup https://trex-tgn.cisco.com/trex/doc/trex_astf_vs_nginx.html#_overview ) If we do we want to do real data plane performance testing Trex is the tool, data plane performance has to be measure in packets per second to have a good understanding on how many 

iperf and others userspace tools, are not real dataplane performance, because they are constrained by the system (you can only generate the traffic that the pod is able to generate) and does not give you an idea con how many packet per second is able to move the system, because offloading and other mechanism optimize it

Reply:
Yes i agree with you regarding the iperf. But Trex brings the need for 2 vNICs. How do we deal with that ? If we can use multus CNI for such testcases, then we can use Trex. Is it possible to manage with just 1 vNIC for Trex ?

Reply:
> Is it possible to manage with just 1 vNIC for Trex ?

it is possible https://trex-tgn.cisco.com/trex/doc/trex_manual.html#_dummy_ports